### ISSUE TITLE
When using a HOSTS groups, also generate reverse lookup?

### ISSUE THREAD

**Issue by cbuijs** [folbricht/routedns#107]
When using a hosts-file with the hosts type groups rule, forward lookups for `A` and `AAAA` work fine, but when doing reverse-lookup for the IP (`PTR`), no answers are given. Could this feature be added? As the ingredients to formulate an answer are already there?

This when using entries in the hosts-file other then "Null" IP-Addresses (`0.0.0.0` and `::`) of course.

Example hosts file:
```
192.168.1.10 chris.home
192.168.1.25 printer.home
```
A reverse lookup for 192.168.1.25 (`25.1.168.192.in-addr.arpa/PTR`) would result in `printer.home` as answer.

**Comment by folbricht** [comment]
This should be doable, but it'll get confusing when dealing with a hosts file that has everything set to 127.0.0.1 or similar. Basically, it'll have to pick the first (or last) from the list at random when the IP is not unique. It can be done though and will work fine if there's a real 1:1 mapping

**Comment by cbuijs** [comment]
It doesn't have to be mutual exclusive or unique, it is fine to answer multiple answers (but you can max it for example to the 10 first ones to not overload things or whatever).

It would look like this:
```
dig @127.0.0.1 -x 127.0.0.1
; <<>> DiG 9.16.8 <<>> @127.0.0.1 -x 127.0.0.1
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 1518
;; flags: qr rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;1.0.0.127.in-addr.arpa.		IN	PTR

;; AUTHORITY SECTION:
1.0.0.127.in-addr.arpa. 12860 IN	PTR	laptop-wifi.home.
1.0.0.127.in-addr.arpa. 12860 IN	PTR	laptop-eth.home.
1.0.0.127.in-addr.arpa. 12860 IN	PTR	test.home.

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Tue Dec 15 08:39:07 CET 2020
;; MSG SIZE  rcvd: 100
```
That said, if the IP is `127.0.0.1`, `::1`, `0.0.0.0` or `::` you could concider to generate a NXDOMAIN anyway (or configurable resolver for particular IPs), so it becomes a mute point. Normally this is used on the public-blocklists anyway.

**Comment by folbricht** [comment]
The `issue-107` branch has a first implementation. For now it just returns one record even if more than one name resolve to the same IP (the last one in the list). I could extend this to a list, up to a max perhaps if necessary.

**Comment by cbuijs** [comment]
Will check it out, returning the first entry should suffice for most applications anyway.

**Comment by cbuijs** [comment]
Works beautifully! Thanks!

**Comment by cbuijs** [comment]
Re-opening this, having a case where a webserver, using the same IP-Address but different domain-names (different web-sites hosted on the same box). Where sometimes, for security checks the reverse has to include the domain-name checked (same apply for mail-servers using/hosting multiple mail-domains on the same box).

Returning all is actually also how most DNS servers act/work.

So returning all PTR's would be a nice option, maybe even configurable (`return = all | first | last`).

**Comment by cbuijs** [comment]
Ping!

**Comment by cbuijs** [comment]
Ping! - Friendly Reminder.

**Comment by folbricht** [comment]
Can you try `issue-107-names`? One thing I'm worried about is a reasonable limit. If there's a large list with everything mapped to 127.0.0.1 then this could cause problems. Any suggestions for a good limit on the number of records in the response?

**Comment by cbuijs** [comment]
It maybe was not explained rightly. I applied the patch, but result remains the same.

When I use this in a hosts file:
```
1.2.3.4 tester.fritz.box.
4.3.2.1 tester.fritz.box.
2001:db8::1 tester.fritz.box.
2001:db8::2 tester.fritz.box.
```
The `dig tester.fritz.box` only returns `4.3.2.1` (as before), and I want it to return all (per IP family) entries.

I think limiting to three per IP-Family (3 for IPv4 and 3 for IPv6, so six in total) would be fine. Maybe add it as a parameter and have this as default?

**Comment by folbricht** [comment]
Can you try with `example-config/blocklist-hosts.toml`? This return the following for me

```
$ dig -p 1153 @127.0.0.1 -x 127.0.0.1

; <<>> DiG 9.18.17 <<>> -p 1153 @127.0.0.1 -x 127.0.0.1
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 49701
;; flags: qr rd; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;1.0.0.127.in-addr.arpa.		IN	PTR

;; ANSWER SECTION:
1.0.0.127.in-addr.arpa.	3600	IN	PTR	www.evil.com.
1.0.0.127.in-addr.arpa.	3600	IN	PTR	www.domain.com.

;; Query time: 1 msec
;; SERVER: 127.0.0.1#1153(127.0.0.1) (UDP)
;; WHEN: Mon Sep 04 14:12:19 CEST 2023
;; MSG SIZE  rcvd: 138
```

Wasn't this the intention? There's another issue asking for all IPs from the list to be returned, but this was only for PTR lookups. The other issue hasn't been addressed yet as this one would have to be merged first to avoid a code-conflict.

**Comment by cbuijs** [comment]
Yes, the PTR works (confirmed!), but we need the A/AAAA as well :-).

My mistake, looking at the history of this issue, I mixed it up. But this (PTR lookup) works now.

Sorry for the confusion :-).

**Comment by folbricht** [comment]
oops, just pushed a fix

**Comment by cbuijs** [comment]
Looks like it works as expected. For me this can be merged/closed.

**Comment by folbricht** [comment]
Merged in https://github.com/folbricht/routedns/pull/115
